### DISCUSSION TITLE
HTTP/2 PING and server side reset

### DISCUSSION THREAD

**Discussion by EndUser509** [mitmproxy/mitmproxy#5044] [General]
I have noticed that for instance https://heise.de, when proxied, returns a '502' HTTP error to the client when I click on a link after not doing anything (=reading the page) for over 60 seconds. What I see on the outgoing interface is that the remote server resets the connection the moment the connection is used again (TCP RST). 
When not proxied I observe that the client sends a HTTP/2 PING after 60 seconds which seems to keep the connection alive. Varying the idle time from 60 to 120 seconds, I never manage to reset the connection and a link always loads. 
I wonder if there is a way to remedy the proxied behavior. I am not familiar enough with HTTP/2 to see if this PING interval is something that client and server agree to. Or if this is 'best practice' for browsers. Any ideas where to start? Would it make sense that the proxy sends regular PINGs for idle HTTP/2 connections?

**Comment by mhils** [comment]
Thanks for the detailed bug description! :cake:

> What I see on the outgoing interface is that the remote server resets the connection the moment the connection is used again (TCP RST).

Ah, such great server behavior. 😄

> Varying the idle time from 60 to 120 seconds, I never manage to reset the connection and a link always loads.

After increasing `network.http.spdy.ping-threshold` to 90 in my Firefox settings, I can reproduce the same problem without mitmproxy in the middle:

![image](https://user-images.githubusercontent.com/1019198/148807768-ffa4fd89-75f4-4a66-81ef-b74d7ab3fc38.png)

My test page is https://www.heise.de/Datenschutzerklaerung-der-Heise-Medien-GmbH-Co-KG-4860.html, clicking the "Datenschutz" link at the bottom of the page after 70s is a no-op with the increased ping timeout. The page only (re)loads after clicking it again.

> I am not familiar enough with HTTP/2 to see if this PING interval is something that client and server agree to.

AFAIK there is nothing formally standardized here, there definitely is no negotiation as part of HTTP/2. Firefox just pings every 58 seconds by default. I presume Chrome does something similar.

> Would it make sense that the proxy sends regular PINGs for idle HTTP/2 connections?

Forwarding HTTP/2 pings would be the obvious thing to do. But we deliberately don't do that as that would create a side channel through which an adversary could get data without it being logged. 

As you already suggested, another way forward here is to ping all remotes every N seconds (irrespective of what the client is doing). Unfortunately this is relatively nontrivial as we need to keep a timer somewhere. Doing that is in the proxy core would directly violate what we try to achieve with sans-io, so we either need to add some timer/alarm/wakeup functionality to our I/O handler, or have an addon that keeps track of open connections an injects a ping every N seconds somehow. I'm leaning towards the timer as we may need that anyways for some protocols (and, thinking long term, UDP).

**Reply by EndUser509** [reply]
> > What I see on the outgoing interface is that the remote server resets the connection the moment the connection is used again (TCP RST).
> 
> Ah, such great server behavior. smile

I have not seen other sites that I regularly visit doing this. Maybe its this and the admin tightened it:

http://nginx.org/en/docs/http/ngx_http_v2_module.html#http2_idle_timeout

> > Varying the idle time from 60 to 120 seconds, I never manage to reset the connection and a link always loads.
> 
> After increasing `network.http.spdy.ping-threshold` to 90 in my Firefox settings, I can reproduce the same problem without mitmproxy in the middle:

Aah, that's where the 60ish seconds come from. Good to know.

> My test page is https://www.heise.de/Datenschutzerklaerung-der-Heise-Medien-GmbH-Co-KG-4860.html, clicking the "Datenschutz" link at the bottom of the page after 70s is a no-op with the increased ping timeout. The page only (re)loads after clicking it again.

That is also the best I could do with what _mitmproxy_ allows me to do from a script. I kill the connection when the server sends a reset which causes exactly the same behavior in the browser. Here is what I do:

```
def error(flow: http.HTTPFlow):
    if flow.error is not None and flow.error.msg == 'peer closed connection':
        ctx.log("Hiding error from client")
        flow.kill()
```

From the user perspective (mine, actually) that is better than the 502, no need to press 'back', just tap/click twice.

> > Would it make sense that the proxy sends regular PINGs for idle HTTP/2 connections?
> 
> Forwarding HTTP/2 pings would be the obvious thing to do. But we deliberately don't do that as that would create a side channel through which an adversary could get data without it being logged.

Maybe it depends on your use case? If I use _mitmdump_ to just record all of my surfing, would that affect me? 

If not, would that approach be cheaper to implement than others below? As a side note: Maybe it would be practical to have a low level tweaks class with just settings that can be configured from the `-s` script, overriding sane defaults that you choose, no safety net provided? Stuff like disabling strict HTTP/2 headers  (#4836, one that I currently have patched in my checkout) or letting PINGs pass? The more complex this gets the more of these border cases will pop up that you do not really want to fully cover but at least give some way to deal with.

> As you already suggested, another way forward here is to ping all remotes every N seconds (irrespective of what the client is doing). Unfortunately this is relatively nontrivial as we need to keep a timer somewhere. Doing that is in the proxy core would directly violate what we try to achieve with sans-io, so we either need to add some timer/alarm/wakeup functionality to our I/O handler, or have an addon that keeps track of open connections an injects a ping every N seconds somehow. I'm leaning towards the timer as we may need that anyways for some protocols (and, thinking long term, UDP).

Maybe loop dealing with the server-facing socket could ask the HTTP layer to provide a PING (if the current state allows for one) if but you are right, the PING in this case is kind of out-of-band from the normal client-server communication. And QUIC/HTTP3 also uses PING so this is going to stay...

**Reply by mhils** [reply]
> From the user perspective (mine, actually) that is better than the 502, no need to press 'back', just tap/click twice.

Yeah, I see where you are coming from. The 502 message simplifies debugging for other use cases, so there's a bit of a conflict - I'd like to keep the current behavior where possible.

> Maybe it depends on your use case? If I use _mitmdump_ to just record all of my surfing, would that affect me?

That absolutely depends on the use case and wouldn't affect you at all. My point here is that mitmproxy needs to record everything that passes through the proxy *by default*. That's a very important detail in a forensic setting - you can't get data through it without it being logged.

> If not, would that approach be cheaper to implement than others below? 

Not sure if it's cheaper to implement. Right now only the respective HTTP/2 connection has knowledge of pings/pongs. For forwarding you would need to pass around ping and pong events. This is additionally complicated by the fact that there's not a 1-on-1 mapping between client connections and server connections. You may have multiple server connections belonging to one client connection because some requests have been redirected. Now you need to think about how you correctly distribute pings, etc.. That's all solvable, but likely creates more complexity than stupidly pinging every 58s. :)

> Maybe it would be practical to have a low level tweaks class with just settings that can be configured from the `-s` script, overriding sane defaults that you choose, no safety net provided? Stuff like disabling strict HTTP/2 headers (#4836, one that I currently have patched in my checkout) or letting PINGs pass?

#4836 is "just" a bug we need to fix, we shouldn't do unnecessary validation by default and just forward stuff. 
Other than that I'm absolutely not opposed to low-level options to control protocol details. For example, #4799 would be a prime candidate for that if we observe it more often.

**Reply by EndUser509** [reply]
> > From the user perspective (mine, actually) that is better than the 502, no need to press 'back', just tap/click twice.
> 
> Yeah, I see where you are coming from. The 502 message simplifies debugging for other use cases, so there's a bit of a conflict - I'd like to keep the current behavior where possible.

No need to change that, with the `error` function everybody can deal with this selectively in a script. Works for me.

I am still hoping for a way to avoid the connection reset. If the double-tap annoys me too much, maybe I will look if I can selectively force connections from sites to HTTP/1 in the ClientConnected() hook and just do that for heise.de. :)

> > Maybe it depends on your use case? If I use _mitmdump_ to just record all of my surfing, would that affect me?
> 
> That absolutely depends on the use case and wouldn't affect you at all. My point here is that mitmproxy needs to record everything that passes through the proxy _by default_. That's a very important detail in a forensic setting - you can't get data through it without it being logged.

Your user group has different uses of the tool that you provide, maybe an add-on that sets defaults optimized for certain use case would be cool? It could interact directly with the tweaks proposed below. Something like `--op-mode [forensic|recording|stealthy|...]`.

> > If not, would that approach be cheaper to implement than others below?
> 
> Not sure if it's cheaper to implement. Right now only the respective HTTP/2 connection has knowledge of pings/pongs. For forwarding you would need to pass around ping and pong events. This is additionally complicated by the fact that there's not a 1-on-1 mapping between client connections and server connections. You may have multiple server connections belonging to one client connection because some requests have been redirected. Now you need to think about how you correctly distribute pings, etc.. That's all solvable, but likely creates more complexity than stupidly pinging every 58s. :)

Maybe this could be a compromise for an initial solution with just the goal to get the PING to the remote server without data leak:

If the client sends a PING, the proxy sends the ACK back immediately. There is nothing that you need to know from the server to do that. What it also triggers is a PING sent to the remote server with a random payload (or the current time as a UINT64 or whatever), best effort, if the highest layer is http2, established, idle and last PING more then n seconds ago. That way no data is leaked to the server. 

The only 'wrong' thing is the round-trip time seen by the client is probably way shorter then with the real server. And you could morse code data to the server by timing PINGs. But then again you could do that with normal requests too. A 'Ping Received' event could be logged on the console so it would not go unnoticed.

I have looked at the code. h2 notifies of a PingReceived, it's just ignored in `http/__init__.py`. If a new 'ForwardPing' command could be passed for processing by the server side connection handler, then a PING could be sent, fire-and-forget. If I understand the RFC, the PING is on stream 0 and therefore can always be sent. h2 has a Ping() function to just do that. 

Unfortunately the whole command handling code in the `http` module eludes me, so I cannot quickly come up with a working sample. :)

> > Maybe it would be practical to have a low level tweaks class with just settings that can be configured from the `-s` script, overriding sane defaults that you choose, no safety net provided? Stuff like disabling strict HTTP/2 headers (#4836, one that I currently have patched in my checkout) or letting PINGs pass?
> 
> #4836 is "just" a bug we need to fix, we shouldn't do unnecessary validation by default and just forward stuff. Other than that I'm absolutely not opposed to low-level options to control protocol details. For example, #4799 would be a prime candidate for that if we observe it more often.

I think you should just add something with one boolean to test the concept, maybe accessible over `ctx`, something global. Once you have that then maybe more contributions with stuff that can be in the code but disabled by default will come in. The good thing is that this way the pre-compiled binary distributions can be used by everybody. Currently adding optional code to the main code for a non core-dev is a pain, one would need to add command line options, documentation, etc. For the tweaks, the documentation could be in the .py file.  

You could also move some of the hard coded defaults to that class, like the listen queue length, or any constant you find in the code that somebody just _might_ want to change. STORE_CAP in certs.py for instance.

**Reply by mhils** [reply]
If you need a stopgap approach I would either force HTTP/1 (you should be able to set `server.alpn` in the `server_connect` hook), or just patch [TimeoutWatchdog.CONNECTION_TIMEOUT](https://github.com/mitmproxy/mitmproxy/blob/1e9707307815144a3106c3b5056111971ad07c6d/mitmproxy/proxy/server.py#L31) to 58 seconds. Much easier than forwarding pings. 😉 

> Unfortunately the whole command handling code in the `http` module eludes me, so I cannot quickly come up with a working sample. :)

Yes, it's not super straightforward unfortunately. If you are interested, here are some pointers:

High-level architecture:

![image](https://user-images.githubusercontent.com/1019198/149019475-d92c28a2-64f5-4e36-88d4-e7cb73b02208.png)
(from https://mitmproxy.org/posts/new-proxy-core/)

The "HTTP Proxy" part in the illustration is the HttpLayer:
https://github.com/mitmproxy/mitmproxy/blob/1e9707307815144a3106c3b5056111971ad07c6d/mitmproxy/proxy/layers/http/__init__.py#L658-L666
Received data ends up in `_handle_event` where it is handled here:
https://github.com/mitmproxy/mitmproxy/blob/1e9707307815144a3106c3b5056111971ad07c6d/mitmproxy/proxy/layers/http/__init__.py#L741-L742
`event_to_child` then passes the data to the HTTP/2 server implementation and distributes all the events returned by that. So this is where you would do the forwarding. But again, I would go for one of the much easier approaches listed above or implement timers right away. Timers aren't that much harder than forwarding: You yield a to-be-defined `RequestWakeup()` when Http2Connection is started [here](https://github.com/mitmproxy/mitmproxy/blob/1e9707307815144a3106c3b5056111971ad07c6d/mitmproxy/proxy/layers/http/_http2.py#L95); react to that in `proxy/server.py` by spawning a new asyncio task/coroutine that first sleeps the defined amount of time and then calls `.server_event(Wakeup(wakeup_request))`. Happy to answer more questions if you are interested, but don't feel obliged to. 😉 

> I think you should just add something with one boolean to test the concept, maybe accessible over `ctx`, something global. Once you have that then maybe more contributions with stuff that can be in the code but disabled by default will come in. The good thing is that this way the pre-compiled binary distributions can be used by everybody. Currently adding optional code to the main code for a non core-dev is a pain, one would need to add command line options, documentation, etc. For the tweaks, the documentation could be in the .py file.

I'd be more than interested to hear how we can make custom optional code easier, but I'm not sure if I understand what you are proposing. Could you clarify?

> You could also move some of the hard coded defaults to that class, like the listen queue length, or any constant you find in the code that somebody just _might_ want to change. STORE_CAP in certs.py for instance.

I'm not following, sorry. Are you proposing we move STORE_CAP somewhere else, or are you proposing we move other hardcoded integers into constants like we did with STORE_CAP?

**Reply by EndUser509** [reply]
> If you need a stopgap approach I would either force HTTP/1 (you should be able to set `server.alpn` in the `server_connect` hook), or just patch [TimeoutWatchdog.CONNECTION_TIMEOUT](https://github.com/mitmproxy/mitmproxy/blob/1e9707307815144a3106c3b5056111971ad07c6d/mitmproxy/proxy/server.py#L31) to 58 seconds. Much easier than forwarding pings. wink

Probably. :)
 
> > Unfortunately the whole command handling code in the `http` module eludes me, so I cannot quickly come up with a working sample. :)
> 
> Yes, it's not super straightforward unfortunately. If you are interested, here are some pointers:
> 
> High-level architecture:
> 
> ![image](https://user-images.githubusercontent.com/1019198/149019475-d92c28a2-64f5-4e36-88d4-e7cb73b02208.png) (from https://mitmproxy.org/posts/new-proxy-core/)
> 
> The "HTTP Proxy" part in the illustration is the HttpLayer:
> 
> https://github.com/mitmproxy/mitmproxy/blob/1e9707307815144a3106c3b5056111971ad07c6d/mitmproxy/proxy/layers/http/__init__.py#L658-L666
> 
> Received data ends up in `_handle_event` where it is handled here:
> 
> https://github.com/mitmproxy/mitmproxy/blob/1e9707307815144a3106c3b5056111971ad07c6d/mitmproxy/proxy/layers/http/__init__.py#L741-L742
> 
> `event_to_child` then passes the data to the HTTP/2 server implementation and distributes all the events returned by that. So this is where you would do the forwarding. But again, I would go for one of the much easier approaches listed above or implement timers right away. Timers aren't that much harder than forwarding: You yield a to-be-defined `RequestWakeup()` when Http2Connection is started [here](https://github.com/mitmproxy/mitmproxy/blob/1e9707307815144a3106c3b5056111971ad07c6d/mitmproxy/proxy/layers/http/_http2.py#L95); react to that in `proxy/server.py` by spawning a new asyncio task/coroutine that first sleeps the defined amount of time and then calls `.server_event(Wakeup(wakeup_request))`. Happy to answer more questions if you are interested, but don't feel obliged to. wink

Thanks for the overview. I still may try to do the PING forward as an exercise. Now I have a clue where to start.

> > I think you should just add something with one boolean to test the concept, maybe accessible over `ctx`, something global. Once you have that then maybe more contributions with stuff that can be in the code but disabled by default will come in. The good thing is that this way the pre-compiled binary distributions can be used by everybody. Currently adding optional code to the main code for a non core-dev is a pain, one would need to add command line options, documentation, etc. For the tweaks, the documentation could be in the .py file.
> 
> I'd be more than interested to hear how we can make custom optional code easier, but I'm not sure if I understand what you are proposing. Could you clarify?
> 
> > You could also move some of the hard coded defaults to that class, like the listen queue length, or any constant you find in the code that somebody just _might_ want to change. STORE_CAP in certs.py for instance.
> 
> I'm not following, sorry. Are you proposing we move STORE_CAP somewhere else, or are you proposing we move other hardcoded integers into constants like we did with STORE_CAP?

I sent a pull-request (not for merging!) that illustrates what I mean. It is non-functional, as the script does not seem to have access to globals in the main program. The example is for a fictional quick-fix for #4836, only to illustrate the point. 

This would need more fleshing out, maybe with separate Defaults(), Globals(), QuickFixes(), etc. classes to group by intention. You can use this in many ways, quick and dirty fixes to win time, staging of features that start like this before they become first-class citizens with proper end-user documentation and command line options, etc. You could maybe compare it to about:config in Firefox.

**Reply by EndUser509** [reply]
https://github.com/mitmproxy/mitmproxy/pull/5051

**Comment by mhils** [comment]
Thanks for the proof-of-concept at https://github.com/mitmproxy/mitmproxy/pull/5051! 😃 This makes more sense now.
It feels like what you propose is actually not too far from the options we have?
For example, we have a `keep_host_header` option defined here:

https://github.com/mitmproxy/mitmproxy/blob/1e9707307815144a3106c3b5056111971ad07c6d/mitmproxy/addons/proxyserver.py#L108-L114

which then is just used in the proxy logic to conditionally update the host header:

https://github.com/mitmproxy/mitmproxy/blob/ef8c88da1f1e7bc15989e285b5565f9540c6d5f1/mitmproxy/proxy/layers/http/__init__.py#L190-L196

You can set that manually (`mitmproxy --set keep_host_header`), specify it `config.yaml`, use the interactive options editor to toggle it while mitmproxy is running, or as in your example write a script that forces it to be a specific value:
```python
from mitmproxy.ctx import options
options.keep_host_header = True
```

Is there something that you are missing from this?

**Reply by EndUser509** [reply]
Nothing missing, the result is the same but the road is a different one. :) 

If that approach is usable kind of depends on your design policies, quality management and openness to some degree of clutter:

1. Do you require proper documentation for the option also in the online documentation or is `-h` enough?
2. How open are you to have tons of additional switches with their respective add-ons, that would also vanish again quickly if they become obsolete? If you consider my fictional example, would that have been accepted as a temporary add-on?
3. In the binary distributions, can you add your own add-ons? Frankly I have never tried, I personally always use the git version directly. There I can obviously.
4. How would you approach a solution that requires custom code to be called? Something where you set an instance of a class or function that is called in the code? Would these be hooks, and if yes, how open are you to allow for custom hooks?
5. Would it work for you if the code evaluating the option is always in but the add-on part is only in an 'examples' folder and you have to pick the ones that you want, provided that 3. works?

The solution with options and documentation is what I consider first class. In a tweaks class you can make it more obvious that the warranty is void and changes are for experts only. That is maybe the essential difference. 

I'm sorry, this ended up in a quite fundamental discussion, off the original topic. :smile:

**Reply by mhils** [reply]
> * Do you require proper documentation for the option also in the online documentation or is `-h` enough?

A docstring as in the `keep_host_header` example above is enough for niche use cases. https://docs.mitmproxy.org/dev/concepts-options/ is just generated from that, so it's always included in the online documentation to some extent.

> * How open are you to have tons of additional switches with their respective add-ons, that would also vanish again quickly if they become obsolete? If you consider my fictional example, would that have been accepted as a temporary add-on?

Generally yes, as long as they may be useful to others, have little maintenance overhead, and no security implications. For your fictional example, I would probably require that the option name includes `insecure` and suggest that the docstring includes something along the lines of "This is a temporary mitigation for issue #xxx and will likely be removed in a future version.". I'm happy to merge provisional stopgap fixes if they are engineered cleanly. It's a unequivocal improvement over the status quo. At the same time, I would probably say no to a `STORE_CAP` option unless you bring up good arguments for why monkeypatching isn't enough.

It obviously also depends on who is asking. I am much more willing to help someone who previously contributed excellent bug reports (#5034) as opposed to the random drive-by code dump. :)

> * In the binary distributions, can you add your own add-ons? Frankly I have never tried, I personally always use the git version directly. There I can obviously.

Adding your own addons is always possible. The binaries won't play nicely if your addon has additional external dependencies, in which case we recommend [pipx](https://docs.mitmproxy.org/dev/overview-installation/#installation-from-the-python-package-index-pypi). Or you use pip/PyPI, which boils down to the same. :)

> * How would you approach a solution that requires custom code to be called? Something where you set an instance of a class or function that is called in the code? Would these be hooks, and if yes, how open are you to allow for custom hooks?

This depends on the specific case. Generally speaking event hooks need to be widely useful and have a much higher bar than options. I'm generally happy to merge slight structural changes which allow monkeypatching for niche use cases. We have a few exotic custom code examples as well, for example https://github.com/mitmproxy/mitmproxy/blob/main/examples/addons/http-stream-modify.py.

> * Would it work for you if the code evaluating the option is always in but the add-on part is only in an 'examples' folder and you have to pick the ones that you want, provided that 3. works?

I'm not sure if I understand that correctly, but we won't expose an option that is a noop unless another addon is loaded. Or maybe I'm misunderstanding things, a specific example would help. :)

**Reply by EndUser509** [reply]
Thanks a lot for your time. I see now that the options approach should work well for most cases. Keep it as it is.

I have two features that I added to a private fork. One is an improvement for an existing add-on and after some brushing up I'll make a pull request. I feel pretty confident that this will make some peoples lives easier.

> This depends on the specific case. Generally speaking event hooks need to be widely useful and have a much higher bar than options. I'm generally happy to merge slight structural changes which allow monkeypatching for niche use cases. We have a few exotic custom code examples as well, for example https://github.com/mitmproxy/mitmproxy/blob/main/examples/addons/http-stream-modify.py.

The second improvement relies on a function being called for processing request and response data. With that link you maybe just gave me the right pointer to do this without changes to core code. I will try to rewrite it as an add-on in the same fashion and propose it via pull request. At least this way hopefully your whole effort here was not for nothing. :smile:

> > * Would it work for you if the code evaluating the option is always in but the add-on part is only in an 'examples' folder and you have to pick the ones that you want, provided that 3. works?
> 
> I'm not sure if I understand that correctly, but we won't expose an option that is a noop unless another addon is loaded. Or maybe I'm misunderstanding things, a specific example would help. :)

Basically in our scenario there are always two parts, the add-on with the option definition and parsing, and the code fragment in the core code that uses the supplied options. The question was if you could have the later only. The actual add-ons would not be all active and for arcane things one would have to move an add-on from an inactive folder to `addons`. That would avoid too many options that confuse the casual users. But it's just a thought, since you are open to new options anyways there may be no need to optimize here.

**Comment by EndUser509** [comment]
I gave it a shot to implement the ping forwarding. I came up with this (can do a PR if this helps):

```diff
diff --git a/mitmproxy/proxy/layers/http/__init__.py b/mitmproxy/proxy/layers/http/__init__.py
index 8e4e5b764..10d3c75c4 100644
--- a/mitmproxy/proxy/layers/http/__init__.py
+++ b/mitmproxy/proxy/layers/http/__init__.py
@@ -16,7 +16,7 @@ from mitmproxy.proxy.layers.http import _upstream_proxy
 from mitmproxy.proxy.utils import expect
 from mitmproxy.utils import human
 from mitmproxy.websocket import WebSocketData
-from ._base import HttpCommand, HttpConnection, ReceiveHttp, StreamId
+from ._base import HttpCommand, HttpConnection, ReceiveHttp, ReceivePing, StreamId
 from ._events import HttpEvent, RequestData, RequestEndOfMessage, RequestHeaders, RequestProtocolError, RequestTrailers, \
     ResponseData, ResponseEndOfMessage, ResponseHeaders, ResponseProtocolError, ResponseTrailers
 from ._hooks import HttpConnectHook, HttpErrorHook, HttpRequestHeadersHook, HttpRequestHook, HttpResponseHeadersHook, \
@@ -765,6 +765,9 @@ class HttpLayer(layer.Layer):
                     yield from self.make_stream(command.event.stream_id)
                 stream = self.streams[command.event.stream_id]
                 yield from self.event_to_child(stream, command.event)
+            elif isinstance(command, ReceivePing):
+                conn = self.connections[command.connection]
+                yield from self.event_to_child(conn, command.event)
             elif isinstance(command, SendHttp):
                 conn = self.connections[command.connection]
                 yield from self.event_to_child(conn, command.event)
diff --git a/mitmproxy/proxy/layers/http/_base.py b/mitmproxy/proxy/layers/http/_base.py
index 5e1946b8b..f95aec772 100644
--- a/mitmproxy/proxy/layers/http/_base.py
+++ b/mitmproxy/proxy/layers/http/_base.py
@@ -28,6 +28,18 @@ class HttpCommand(commands.Command):
     pass
 
 
+class ReceivePing(HttpCommand):
+    event: events.Event
+    connection: Connection
+
+    def __init__(self, event: events.Event, conn: Connection):
+        self.event = event
+        self.connection = conn
+
+    def __repr__(self) -> str:
+        return f"ReceivePing({self.event})"
+
+
 class ReceiveHttp(HttpCommand):
     event: HttpEvent
 
diff --git a/mitmproxy/proxy/layers/http/_events.py b/mitmproxy/proxy/layers/http/_events.py
index edddf5b6e..1d4205934 100644
--- a/mitmproxy/proxy/layers/http/_events.py
+++ b/mitmproxy/proxy/layers/http/_events.py
@@ -1,6 +1,7 @@
 from dataclasses import dataclass
 from typing import Optional
 
+from mitmproxy.proxy import events
 from mitmproxy import http
 from mitmproxy.http import HTTPFlow
 from ._base import HttpEvent
@@ -97,6 +98,14 @@ class ResponseProtocolError(HttpEvent):
         self.code = code
 
 
+@dataclass
+class ForwardPing(events.Event):
+    ping_data: str
+
+    def __init__(self, ping_data: str):
+        self.ping_data = ping_data
+
+
 __all__ = [
     "HttpEvent",
     "RequestHeaders",
@@ -109,4 +118,5 @@ __all__ = [
     "ResponseEndOfMessage",
     "RequestProtocolError",
     "ResponseProtocolError",
+    "ForwardPing",
 ]
diff --git a/mitmproxy/proxy/layers/http/_http2.py b/mitmproxy/proxy/layers/http/_http2.py
index 8e477c34f..75a0da6c5 100644
--- a/mitmproxy/proxy/layers/http/_http2.py
+++ b/mitmproxy/proxy/layers/http/_http2.py
@@ -16,9 +16,9 @@ from mitmproxy import http, version
 from mitmproxy.connection import Connection
 from mitmproxy.net.http import status_codes, url
 from mitmproxy.utils import human
-from . import RequestData, RequestEndOfMessage, RequestHeaders, RequestProtocolError, ResponseData, \
-    ResponseEndOfMessage, ResponseHeaders, RequestTrailers, ResponseTrailers, ResponseProtocolError
-from ._base import HttpConnection, HttpEvent, ReceiveHttp, format_error
+from ._events import RequestData, RequestEndOfMessage, RequestHeaders, RequestProtocolError, ResponseData, \
+    ResponseEndOfMessage, ResponseHeaders, RequestTrailers, ResponseTrailers, ResponseProtocolError, ForwardPing
+from ._base import HttpConnection, HttpEvent, ReceiveHttp, ReceivePing, format_error
 from ._http_h2 import BufferedH2Connection, H2ConnectionLogger
 from ...commands import CloseConnection, Log, SendData
 from ...context import Context
@@ -164,6 +164,8 @@ class Http2Connection(HttpConnection):
 
         elif isinstance(event, ConnectionClosed):
             yield from self.close_connection("peer closed connection")
+        elif isinstance(event, ForwardPing):
+            yield Log("PPP Ping forwarded to the wrong connection: " + repr(event.ping_data) + " -> " + repr(self.conn.address))
         else:
             raise AssertionError(f"Unexpected event: {event!r}")
 
@@ -326,6 +328,10 @@ class Http2Server(Http2Connection):
             self.streams[event.stream_id] = StreamState.HEADERS_RECEIVED
             yield ReceiveHttp(RequestHeaders(event.stream_id, request, end_stream=bool(event.stream_ended)))
             return False
+        elif isinstance(event, h2.events.PingReceived):
+            yield Log("PPP Ping received from client: " + repr(event.ping_data))
+            yield ReceivePing(ForwardPing(event.ping_data), self.conn)
+            return False
         else:
             return (yield from super().handle_h2_event(event))
 
@@ -396,7 +402,13 @@ class Http2Client(Http2Connection):
                 yield from self._handle_event(event)
 
     def _handle_event2(self, event: Event) -> CommandGenerator[None]:
-        if isinstance(event, RequestHeaders):
+        if isinstance(event, ForwardPing):
+            yield Log("PPP Ping forwarded: " + repr(event.ping_data) + " -> " + repr(self.conn.address))
+            self.h2_conn.ping(event.ping_data);
+            data_to_send = self.h2_conn.data_to_send()
+            if data_to_send:
+                yield SendData(self.conn, data_to_send)
+        elif isinstance(event, RequestHeaders):
             pseudo_headers = [
                 (b':method', event.request.data.method),
                 (b':scheme', event.request.data.scheme),
@@ -452,6 +464,9 @@ class Http2Client(Http2Connection):
         elif isinstance(event, h2.events.RequestReceived):
             yield from self.protocol_error(f"HTTP/2 protocol error: received request from server")
             return True
+        elif isinstance(event, h2.events.PingAckReceived):
+            yield Log("PPP Ping Ack received from server: " + repr(event.ping_data))
+            return False
         elif isinstance(event, h2.events.RemoteSettingsChanged):
             # We have received at least one settings from now,
             # which means we can rely on the max concurrency in remote_settings
```
This does not work as expected, when a ping is received it is sent back to the client instead of the server. I get log entries like this:

```
192.168.0.2:39850: PPP Ping received from client: b'\x00\x00\x00\x00\x00\x00\x00\x00'
192.168.0.2:39850: PPP Ping forwarded to the wrong connection: b'\x00\x00\x00\x00\x00\x00\x00\x00' -> ('192.168.0.2', 39850)
```
Can you give me a pointer how I can manage to forward the command/event to the right instance of `Http2Client`? If I get this to work I would propose a clean PR (with better names for command/event and other suggestions).

**Reply by mhils** [reply]
> Can you give me a pointer how I can manage to forward the command/event to the right instance of Http2Client?

```diff
diff --git a/mitmproxy/proxy/layers/http/__init__.py b/mitmproxy/proxy/layers/http/__init__.py
index 8e4e5b764..10d3c75c4 100644
--- a/mitmproxy/proxy/layers/http/__init__.py
+++ b/mitmproxy/proxy/layers/http/__init__.py
@@ -765,6 +765,9 @@ class HttpLayer(layer.Layer):
                     yield from self.make_stream(command.event.stream_id)
                 stream = self.streams[command.event.stream_id]
                 yield from self.event_to_child(stream, command.event)
+            elif isinstance(command, ReceivePing):
+                conn = self.connections[command.connection]
+                yield from self.event_to_child(conn, command.event)
```

here you are sending the ReceivePing right back to the connection where it came from. Instead, you probably need to go through
https://github.com/mitmproxy/mitmproxy/blob/83b43b93a7ac6cf2bfe557964fd63b376e158e25/mitmproxy/proxy/layers/http/__init__.py#L674

and pick the appropriate one(s). But please don't stress yourself about a PR, I really think we want a timer-based approach in the end. That is, unless we find that your iPad does indeed have better tricks. :)

**Comment by EndUser509** [comment]
BTW, from my iPad I do not receive any PINGs but the behavior is the same, when proxied I get a 502 after 60 seconds of idle time. I guess Safari uses another HTTP/2 management primitive to keep a connection alive. I'll try to figure out which one it is.

**Reply by mhils** [reply]
That would be interesting, yes please! :)

**Reply by EndUser509** [reply]
I think I figured it out. When I click on a link on my iPad after 60 seconds or so, Safari sends a PRIORITY frame for probing. I can see that with some more `yield Log()` lines added to `_http2.py`. If the remote server has already closed the connection this will result in a TCP RST being returned. Safari then proceeds to just open a new connection. For shorter delays I can see in tcpdump that it again sends one TLS packet (the PRIORITY frame), gets a TCP ACK for it and then continues to use the still alive HTTP/2 streams.

Since the TCP ACK is automatically sent by the TCP stack for receiving the PRIORITY frame, there is no way to delay that and first send one to the remote server to see if you get an ACK or RST back and pass that to the client. That would definitely be an argument for the PING timer approach that you propose. So I will give up on the forwarding then. :/

**Reply by mhils** [reply]
What a fascinating approach. Thanks for digging into this! 🍰

**Comment by makover11** [comment]
Building web tools requires careful handling of requests, performance, and user experience. Testing different inputs is especially important for applications that [process conversions](https://justmorsecodetranslator.com/
) or structured data.
